Support hosting WebAssembly demo page locally

ChinYikMing opened the pull request:
To enhance the user experience of using WebAssembly version of rv32emu, we shall provide a Makefile target (serve-wasm) to serve the WebAssembly demo page in an easy way.

Since rv32emu leverages TCO (tail call optimization) when being compiled, so we have to tell if the user's browser supports TCO. Currently, only Chrome with MAJOR 112 and Firefox with MAJOR 121 supports TCO execution runtime, thus we focus on this two browsers. Also, use color text to remind the users if they need to update their Chrome or Firefox.

See also: #75

ChinYikMing committed:
Support hosting WebAssembly demo page locally

To enhance the user experience of using WebAssembly version of rv32emu,
we shall provide a Makefile target (serve-wasm) to serve the WebAssembly
demo page in an easy way.

Since rv32emu leverages TCO (tail call optimization) when being compiled,
so we have to tell if the user's browser supports TCO. Currently, only
Chrome with MAJOR 112 and Firefox with MAJOR 121 supports TCO execution
runtime, thus we focus on this two browsers. Also, use color text to
remind the users if they need to update their Chrome or Firefox.

See also: #75

jserv commented on mk/color.mk:
```diff
@@ -0,0 +1,3 @@
+YELLOW=\033[0;33m
```
It is not worthy to create a new file containing only 3 lines. Move these to `mk/common.mk`.

jserv reviewed (changes requested):
Move WebAssembly related rules to `mk/wasm.mk`.

ChinYikMing commented:
Run `make CC=emcc start-web` to test it.

Note that you should install emcc then source your emcc firstly. In macOS or Linux, use `source ~/emsdk/emsdk_env.sh`.

ChinYikMing commented:
> Move WebAssembly related rules to `mk/wasm.mk`.

Done.

ChinYikMing commented:
README will be updated on future PR.

ChinYikMing committed:
Move WebAssembly related rules to wasm.mk

mk/external.mk should comes before mk/wasm.mk since mk/wasm.mk depends
on targets which defined in mk/external.mk to fetch externals.

jserv commented on mk/wasm.mk:
```diff
@@ -0,0 +1,104 @@
+STATIC_WEB_FILES := $(WEB_HTML_RESOURCES)/index.html \
+		    $(WEB_JS_RESOURCES)/coi-serviceworker.min.js
+
+serve-wasm: $(BIN) check-demo-dir-exist
```
Rename the target to `start-web`.

ChinYikMing committed:
Rename web target and move colors

jserv reviewed (changes requested):
Update `README.md` to reflect the proposed changes.

ChinYikMing committed:
Update README for WebAssembly Translation

ChinYikMing commented:
> Update `README.md` to reflect the proposed changes.

Done.

jserv commented on README.md:
```diff
@@ -281,6 +281,33 @@ To analyze the profiling data, use the `rv_profiler` tool with the desired optio
+
+### Index page
+You would see a dropdown menu which you can use to select the ELF executable. Select one and
+click the Run button to run it.
```
Mention the demo page as the alternative solution.

jserv commented on README.md:
```diff
@@ -281,6 +281,33 @@ To analyze the profiling data, use the `rv_profiler` tool with the desired optio
 
+## WebAssembly Translation
+### Build and run
+`rv32emu` relies on Emscripten to be compiled to WebAssembly. Thus, the target system should have
```
Put the hyperlink to Emscripten in the first sentence.

jserv commented:
Thank @ChinYikMing for contributing!
